### PR TITLE
Simplifications around command log

### PR DESCRIPTION
**Pull request by terzerm**

### CODE REVIEW

File: src/main/java/org/tools4j/hoverraft/command/CommandLog.java [hunk 1]
```diff
@@ -27,14 +27,15 @@
...
-    void readIndex(long index);
-    LogEntry read(DirectFactory directFactory);
-    void readTo(LogEntry logEntry);
-    int readTerm();
+    LogEntry read(long position, DirectFactory directFactory);
+    void readTo(long position, LogEntry logEntry);
+    void readKeyTo(long position, LogKey logKey);
+    int readTerm(long position);
     void append(LogEntry logEntry);
     void truncateIncluding(long index);
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/CommandLog.java:37]: truncateIncluding(long position)? 
In fact position and logKey.index are kind of redundant and should always be the same

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 2]
```diff
@@ -28,56 +28,43 @@
...
-            readIndex.set((int)index);
-        } else {
-            throw new IllegalArgumentException("Invalid read index " + index + " for log entry with size " + entries.size());
-        }
-    }
-
-    @Override
-    public synchronized int readTerm() {
-        return read(null).logKey().term();
+    public synchronized LogEntry read(final long position, final DirectFactory directFactory) {
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: why directFactory is a parameter? can't it be initialised at construction of the CommandLog?

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 3]
```diff
@@ -87,23 +74,28 @@ public synchronized void append(final LogEntry logEntry) {
...
-        }
     }
 
     @Override
-    public LogKey lastKey() {
-        readIndex(entries.size() - 1);
-        readTo(lastEntry);
-        return lastEntry.logKey();
+    public long lastIndex() {
+        return read(entries.size() - 1).logKey().index();
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: entries.size() - 1 == read(entries.size() - 1).logKey().index()

File: src/test/java/org/tools4j/hoverraft/command/DirectLogEntryTest.java [hunk 4]
```diff
@@ -1,3 +1,26 @@
...
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
```
Comment (anton-anufriev) [inline src/test/java/org/tools4j/hoverraft/command/DirectLogEntryTest.java:23]: I missed it again :|

File: src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java [hunk 5]
```diff
@@ -117,8 +119,7 @@ private void onVoteRequest(final Role currentRole, final int previouslyVotedFor)
 
         when(serverContext.connections().serverSender(candidateId)).thenReturn(sender);
         when(persistentState.votedFor()).thenReturn(previouslyVotedFor);
-        when(persistentState.commandLog().lastKey().term()).thenReturn(lastLogTerm);
-        when(persistentState.commandLog().lastKey().index()).thenReturn(lastLogIndex);
```
Comment (anton-anufriev) [inline src/test/java/org/tools4j/hoverraft/server/VoteRequestHandlerTest.java:121]: nice!

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 6]
```diff
@@ -28,56 +28,43 @@
...
     @Override
     public long size() {
         return entries.size();
     }
 
     @Override
-    public long readIndex() {
-        return readIndex.get();
+    public synchronized int readTerm(final long position) {
+        return read(position, null).logKey().term();
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: why null?

Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: I see. I think directFactory should not be passed as a parameter

**Comment by anton-anufriev** [comment]
Do you want to merge the PR?

**Comment by terzerm** [comment]
Hi Anton,

I have just pushed another bunch of changes. 

I believe most of them make sense. Please have a look.

Cheers,
Marco

File: src/main/java/org/tools4j/hoverraft/command/CommandKeyComparator.java [hunk 7]
```diff
@@ -21,15 +21,16 @@
...
-    long readIndex();
-    void readIndex(long index);
-    M read();
-    void append(M message);
+public enum CommandKeyComparator implements Comparator<CommandKey> {
+    INSTANCE;
+
+    @Override
+    public int compare(final CommandKey commandKey, final CommandKey otherCommandKey) {
+        final int termCompare = Integer.compare(commandKey.sourceId(), otherCommandKey.sourceId());
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/CommandKeyComparator.java]: sourceIdCompare

File: src/main/java/org/tools4j/hoverraft/command/CommandKeyComparator.java [hunk 8]
```diff
@@ -21,15 +21,16 @@
...
+        return termCompare == 0 ? Long.compare(commandKey.commandIndex(), otherCommandKey.commandIndex()) : termCompare;
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/CommandKeyComparator.java]: sourceIdCompare == 0 ... : sourceIdCompare

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 9]
```diff
@@ -32,76 +33,88 @@
 public class InMemoryCommandLog implements CommandLog {
 
     private final DirectFactory directFactory = new AllocatingDirectFactory();
+
     private final List<LogEntry> entries = new ArrayList<>();
+    private final Long2LongHashMap maxCommandIndexBySourceId = new Long2LongHashMap(-1);
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: nice!

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 10]
```diff
@@ -32,76 +33,88 @@
...
-    public synchronized LogEntry read(final long position, final DirectFactory directFactory) {
-        return clone(read(position));
+    public synchronized int readTerm(final long index) {
+        return read(index).logKey().term();
     }
 
     @Override
-    public void readTo(final long position, final LogEntry logEntry) {
-        logEntry.copyFrom(read(position));
+    public synchronized void readKeyTo(final long index, final LogKey logKey) {
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: in the interface: LogKey target
shouldn't they be consistent? or targetLogKey for clarity

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 11]
```diff
@@ -32,76 +33,88 @@
...
     }
 
     @Override
-    public void readKeyTo(final long position, final LogKey logKey) {
-        final LogKey readKey = read(position).logKey();
-        logKey.term(readKey.term()).index(readKey.index());
+    public synchronized void readTo(final long index, final LogEntry target) {
+        final LogEntry source = read(index);
+        target.writeBufferOrNull()
+                .putBytes(0, source.readBufferOrNull(), source.offset(), source.byteLength());
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: target.copyFrom(source)?

Comment (anton-anufriev) [reply src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: ahh, I see, you have removed it.

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 12]
```diff
@@ -32,76 +33,88 @@
...
 
     @Override
-    public synchronized void append(final LogEntry logEntry) {
-        entries.add(clone(logEntry));
+    public void append(final int term, final Command command) {
+        final LogEntry logEntry = directFactory.logEntry();
+        logEntry.logKey().term(term);
+        logEntry.command()
+                .writeBufferOrNull()
+                .putBytes(0, command.readBufferOrNull(), command.offset(), command.byteLength());
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: does command.byteLength() include the byteLength field size (4 bytes)?

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 13]
```diff
@@ -32,76 +33,88 @@
...
+        synchronized (this) {
+            final long index = entries.size();
+            logEntry.logKey().index(index);//TODO this is redundant, do we need this?
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java:81]: I would leave it for now. I thought it was nice to pass prevLogKey to appendRequest and lastLogKey in the voteRequest and apply compare and containment operation on them. But may be it is redundant.

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 14]
```diff
@@ -32,76 +33,88 @@
...
     @Override
     public synchronized void truncateIncluding(final long index) {
-        if (index >= entries.size()) {
-            throw new IllegalArgumentException("Truncate index " + index + " must be less than the size " + entries.size());
-        }
-        for (long idx = lastIndex(); idx >= index; idx--) {
-            entries.remove((int)idx);
+        for (int idx = entries.size() - 1; idx >= index; idx--) {
+            final LogEntry removed = entries.remove(idx);
+            maxCommandIndexBySourceId.remove(removed.command().commandKey().sourceId());
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: it does not look right to me. If you truncate the log there should be other commands from the SourceId left in the log, however the SourceId keyed entry would be removed.

File: src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java [hunk 15]
```diff
@@ -32,76 +33,88 @@
...
+        long maxCommandIndex = maxCommandIndexBySourceId.get(sourceId);
+        if (maxCommandIndex < 0) {
+            for (int idx = entries.size() - 1; idx >= 0; idx--) {
+                final CommandKey ckey = entries.get(idx).command().commandKey();
+                if (ckey.sourceId() == sourceId) {
+                    maxCommandIndex = ckey.commandIndex();
+                    maxCommandIndexBySourceId.put(sourceId, maxCommandIndex);
+                    break;
+                }
+            }
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java]: I see that you are rebuilding the command index.
Is it a right place in contains() method? Shouldn't it be part of truncate method?

File: src/main/java/org/tools4j/hoverraft/command/machine/StateMachine.java [hunk 16]
```diff
@@ -23,9 +23,9 @@
  */
 package org.tools4j.hoverraft.command.machine;
 
-import org.tools4j.hoverraft.message.CommandMessage;
+import org.tools4j.hoverraft.command.Command;
 
 public interface StateMachine {
-    //Should it be bare Command or CommandMessage
-    void onMessage(CommandMessage message);
+    //Should it be bare CommandPayload or Command
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/command/machine/StateMachine.java]: remove comment as we know that it is the command that is applied the the state machine

File: src/main/java/org/tools4j/hoverraft/direct/AllocatingDirectFactory.java [hunk 17]
```diff
@@ -72,9 +73,9 @@ public TimeoutNow timeoutNow() {
     }
 
     @Override
-    public CommandMessage commandMessage() {
-        final DirectCommandMessage directCommandMessage = new DirectCommandMessage();
-        directCommandMessage.wrap(newBuffer(DirectCommandMessage.EMPTY_COMMAND_BYTE_LENGTH), 0);
+    public Command commandMessage() {
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/direct/AllocatingDirectFactory.java]: command() ?

File: src/main/java/org/tools4j/hoverraft/direct/AllocatingDirectFactory.java [hunk 18]
```diff
@@ -72,9 +73,9 @@ public TimeoutNow timeoutNow() {
...
+        final DirectCommand directCommandMessage = new DirectCommand();
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/direct/AllocatingDirectFactory.java]: directCommand

File: src/main/java/org/tools4j/hoverraft/direct/DirectFactory.java [hunk 19]
```diff
@@ -41,7 +42,7 @@
 
     TimeoutNow timeoutNow();
 
-    CommandMessage commandMessage();
+    Command commandMessage();
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/direct/DirectFactory.java]: command()

File: src/main/java/org/tools4j/hoverraft/direct/RecyclingDirectFactory.java [hunk 20]
```diff
@@ -67,8 +68,8 @@ public TimeoutNow timeoutNow() {
     }
 
     @Override
-    public CommandMessage commandMessage() {
-        return commandMessage;
+    public Command commandMessage() {
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/direct/RecyclingDirectFactory.java]: command()

**Comment by terzerm** [comment]
Will add copyFrom back

**Comment by terzerm** [comment]
good point — bug — fixed.

**Comment by terzerm** [comment]
donnow. don’t like this redundancy stuff. it confused me completely.

**Comment by terzerm** [comment]
yea rebuilding index later

**Comment by terzerm** [comment]
you are right. would be nicer to rebuild on truncate.

will change.

**Comment by terzerm** [comment]
ok will remove

**Comment by terzerm** [comment]
yep

**Comment by terzerm** [comment]
renamed

**Comment by terzerm** [comment]
actually i did it this way cause then this also works if we open a (possibly huge) command log at the start.

File: src/main/java/org/tools4j/hoverraft/message/direct/DirectCommandPayload.java [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+public class DirectCommandPayload extends AbstractDirectPayload implements CommandPayload {
+
+    private static final int BYTE_LENGTH_OFF  = 0;
+    private static final int BYTE_LENGTH_LEN  = 4;
+    private static final int BYTES_OFF = BYTE_LENGTH_OFF + BYTE_LENGTH_LEN;
+
+    @Override
+    public int byteLength() {
+        return readBuffer.getInt(offset + BYTE_LENGTH_OFF);
```
Comment (anton-anufriev) [inline src/main/java/org/tools4j/hoverraft/message/direct/DirectCommandPayload.java:48]: I think here is the catch. DirectCommandPayload.byteLength() does not include the bytelength field size.

**Comment by terzerm** [comment]
already fixed by now

**Pull request merged**